### PR TITLE
Add support for parent call function syntax

### PR DESCRIPTION
**Pull request by matt-sheets**
Users will now be able to use the foo\<bar\>.read() syntax

### CODE REVIEW

**Comment by matt-sheets** [comment]
As of posting this draft, we are doing this the "old" way of <parent> instead of ::parent.  But the rest of the logic should be sound.

File: src/internal_rep.rs
Comment (dburgener) [inline src/internal_rep.rs:2326]: This block is pretty hard to follow with all the indentation.  It looks like this whole match block could get pulled out into its own function (returning `Result<(), CascadeErrors>`), which would help readability a lot

File: src/internal_rep.rs
Comment (dburgener) [inline src/internal_rep.rs:2346]: I'm not really sure what the history of this comment is, but it sure seems irrelevant buried down in this block.  I'm not sure it's much better moved outside the block.  It can probably just be deleted, while you're rearranging things here.

File: data/policies/parent_call.cas
Comment (dburgener) [inline data/policies/parent_call.cas:8]: This line and line 14 have spaces, while the rest of the rest of the file has tabs.

File: data/error_policies/parent_call.cas
Comment (dburgener) [inline data/error_policies/parent_call.cas:8]: This line and line 14 have spaces, while the rest of the rest of the file has tabs.

**Comment by dburgener** [comment]
Posted some initial comments.  As you mentioned, the syntax here is likely to change.  We'll try to get a final decision on that soon.

File: src/parser.lalrpop
Comment (matt-sheets) [inline src/parser.lalrpop:133]: This is not going to be the final solution.  Do not merge until this is fixed.

File: data/expected_cil/parent_call.cil
Comment (dburgener) [inline data/expected_cil/parent_call.cil:143]: Discussed on call.  I think this should probably be "(call bar-read (foo dom))" instead.

File: src/functions.rs
Comment (dburgener) [inline src/functions.rs:2075]: There's a top level get_cil_name() in ast.rs that does this logic.

File: src/functions.rs
Comment (dburgener) [inline src/functions.rs:2120]: You should be able to write this more cleanly as a match statement I think.  Something like:

```
let mut args = match (call.class_name, call.cast_name) {
    (None, Some(cast_name)) => { ... },
    (None, None) => { ... },
    (Some(class_name), _) => { ... },
}
```

Comment (dburgener) [reply src/functions.rs:2120]: Of course, rewriting it that way really suggests a further rewrite to:

```
let mut args = match call.class_name {
    Some(class_name) => { ... },
    None => { match call.cast_name {
        Some(cast_name) => { ... },
        None => { ... },
    }
}
```

which might make the intent cleaner.  I'm not sure I've actually followed the logic of this part yet though.

File: data/policies/parent_call.cas
Comment (dburgener) [inline data/policies/parent_call.cas:29]: Are there TODOs in the code for what's left to do functionality wise to support these use cases?

**Review by dburgener — CHANGES_REQUESTED** [review]

File: src/functions.rs
Comment (matt-sheets) [reply src/functions.rs:1811]: This was actually with the old parsing.  I am going to delete this comment for now since I don't think it is accurate anymore.

File: data/policies/parent_call.cas
Comment (matt-sheets) [reply data/policies/parent_call.cas:29]: The TODO added in one of your previous comments should cover both of these.

Comment (matt-sheets) [reply data/policies/parent_call.cas:29]: But this does bring up another question.  Do we want <bar>.read(this); to work with this PR?  Currently this is broken at the parser level.

File: src/functions.rs
Comment (matt-sheets) [reply src/functions.rs:1794]: cast_ti is also None for the validate_argcast stuff, since it was not originally used.  So right now, being None is okay.

File: data/policies/parent_call.cas
Comment (dburgener) [reply data/policies/parent_call.cas]: What do you mean by ".read(this)"?  Not sure I follow the question.

Comment (matt-sheets) [reply data/policies/parent_call.cas]: Sorry markup ate the comment see edit above

Comment (dburgener) [reply data/policies/parent_call.cas]: That makes more sense, but I'm still not sure I follow.  What behavior is `<bar>.read(this);` supposed to enable?  In the test case it looks like it should just be `xyz.read(this);` without brackets?  What's the difference between `<bar>.read(this);` and `bar.read(this)`?

File: src/functions.rs
Comment (dburgener) [reply src/functions.rs]: I think we should pass cast_ti through validate_argcast from validate_argument.  It looks like validate_argument checks the validity of the cast target directly.  That logic should probably move into validate_cast and apply to both situations?

Comment (dburgener) [reply src/functions.rs]: It looks like all we're using the TypeInstance for is to check if it matches TypeInstance::SEType.  That should be equivalent to calling is_setype() on the TypeInfo, I think?

Ideally, casting should work without any need to rely on TypeInstances, since we care what these things *are* not what they're *called*.

File: data/error_policies/parent_call.cas
Comment (dburgener) [inline data/error_policies/parent_call.cas:9]: Doesn't overly matter to the point of the test, but the write() function should probably grant the write permission?

File: src/ast.rs
Comment (dburgener) [inline src/ast.rs:433]: This "if let" followed by handling for the None case really should be a match block.

File: src/functions.rs
Comment (dburgener) [inline src/functions.rs:1776]: Are you sure you're removing the correct return?  The one on 1828 does look needless to me.

File: src/functions.rs
Comment (dburgener) [reply src/functions.rs]: Looking at the latest version of the code, I think the question about cast_ti being None still stands.  The validate_argcast case has become guaranteed to be Some(), and the ValidatedCall case is the return of types.get().

Without even thinking about callers though, I'm not sure how doing any casting when the target is None makes sense, so it should probably be an error.

File: src/functions.rs
Comment (dburgener) [inline src/functions.rs:1849]: It might be a domain too, right?  Probably should say "type does not..."?

File: src/functions.rs
Comment (dburgener) [inline src/functions.rs:2110]: Could this cause a problem if the function name contains the class name as a substring?  In other words, if I have a resource named foo, with a function named foobar().  Then the cil name is foo-foobar(), and if I cast to baz, this replace, would turn it into the invalid baz-bazbar() instead of the intended baz-foobar().

File: src/functions.rs
Comment (dburgener) [inline src/functions.rs:2158]: I feel like this section of code is a little inconsistent about what it assumes about the broader codebase.  I *think* that based on the current state of ast.rs it's impossible at the parser level for class_name to be None while cast_name is Some().  The block starting on 2141 seems to assume that, by not handling the class_name None case, which is reasonable enough.  But now we have special handling for is class_name is None and cast_name is Some(), which I justified the above block as being fine because that's impossible.

I think it is impossible for class_name to be None while cast_name is Some(), but I think this function should either 1. Assume that and act accordingly or 2. Verify that and return an internal error if it's false.   If 1, that should probably be documented in a comment.

File: src/functions.rs
Comment (matt-sheets) [reply src/functions.rs]: The None case for cast_ti should be handled at line 1824 in functions.rs.

**Review by dburgener — APPROVED** [review]

**Comment by dburgener** [comment]
Merged, thanks!

**Pull request merged**
